test/integration/scheduler_perf/scheduler_perf_test.go

```diff
@@ -343,24 +351,24 @@ func runWorkload(b *testing.B, tc *testCase, w *workload) []DataItem {
 			} else {
 				namespace = fmt.Sprintf("namespace-%d", opIndex)
 			}
-			var stopCh chan struct{}
 			var collectors []testDataCollector
+			collectorCtx, collectorCancel := context.WithCancel(workloadCtx)
 			if concreteOp.CollectMetrics {
-				stopCh = make(chan struct{})
 				collectors = getTestDataCollectors(podInformer, fmt.Sprintf("%s/%s", b.Name(), namespace), namespace, tc.MetricsCollectorConfig)
 				for _, collector := range collectors {
-					go collector.run(stopCh)
+					go collector.run(collectorCtx)
 				}
 			}
 			if err := createPods(namespace, concreteOp, clientset); err != nil {
-				b.Fatal(err)
+				b.Fatalf("op %d: %v", opIndex, err)
 			}
-			barrierAndCollect := func() {
-				if err := barrierOne(podInformer, b.Name(), namespace, concreteOp.CreatePods.(int)); err != nil {
-					b.Fatal(err)
+			barrierAndCollect := func(idx int) {
+				if err := waitUntilPodsScheduledInNamespace(podInformer, b.Name(), namespace, concreteOp.CreatePods.(int)); err != nil {
+					barrierAndCollectErrors[idx] = err
```

Comment:
you could still use a channel with just one element and use `fmt.Errorf` to indicate the operation index.

See https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/internal/parallelize/error_channel.go
(but don't use that directly as we prefer to keep that internal)

Reply:
Nice, we only care about one error, so we can drop errors after the first one without blocking.